src/messages_queue.h

```diff
@@ -0,0 +1,272 @@
+    if (new_current_file->fail()) {
+      // If file can't be created, fall back to the in-memory storage.
+      storage_directory_.clear();
+      // TODO(AlexZ): DebugLog it!
+      delete new_current_file;
+    } else {
+      storage_directory_ = directory;
+      current_file_.reset(new_current_file);
+      // Also check if there are any messages in the memory storage, and save them
+      // to file.
+      if (!messages_storage_.empty()) {
+        StoreMessages(messages_storage_);
+        messages_storage_.clear();
+      }
+    }
+  }
+
+  void ProcessMessageCommand() {
+    std::string messages_buffer_copy;
+    {
+      std::lock_guard<std::mutex> lock(mutex_);
+      if (!messages_buffer_.empty()) {
+        messages_buffer_copy.swap(messages_buffer_);
+      }
+    }
+    if (!messages_buffer_copy.empty()) {
+      StoreMessages(messages_buffer_copy);
+    }
+  }
+
+  // If there is no file storage directory set, it should also process messages
+  // from the memory buffer.
+  void ProcessArchivedFilesCommand(TArchivedFileProcessor processor, TFileProcessingFinishedCallback callback) {
+    ProcessingResult result = ProcessingResult::ENothingToProcess;
+    // Process in-memory messages, if any.
+    // TODO(AlexZ): Append user ID to every bunch of data sent to processor.
+    if (!messages_storage_.empty()) {
+      // TODO(AlexZ): gzip it before processing.
+      if (processor(false /* in-memory buffer */, messages_storage_)) {
+        messages_storage_.clear();
+        result = ProcessingResult::EProcessedSuccessfully;
+      } else {
+        result = ProcessingResult::EProcessingError;
+      }
+      // If in-memory storage is used, then file storage directory was not set
+      // and we can't process files.
+      // So here we notify callback and return.
+      // TODO(AlexZ): Do we need to use in-memory storage if storage is
+      // initialized but full/not accessible?
+      if (callback) {
+        callback(result);
+      }
+      return;
+    }
+    if (current_file_ && current_file_->tellp() > 0) {
+      ArchiveCurrentFile();
+    }
+    FileManager::ForEachFileInDir(storage_directory_, [&processor, &result](const std::string & full_path_to_file) {
+      // Ignore non-archived files.
+      if (full_path_to_file.find(kArchivedFilesExtension) == std::string::npos) {
+        return true;
+      }
+      if (processor(true /* file path */, full_path_to_file)) {
+        result = ProcessingResult::EProcessedSuccessfully;
+        // Also delete successfully processed archive.
+        std::remove(full_path_to_file.c_str());
+        return true;
+      } else {
+        result = ProcessingResult::EProcessingError;
+        // Stop processing archives on error.
+        return false;
+      }
+    });
+    if (callback) {
+      callback(result);
+    }
+  }
+
+  void WorkerThread() {
+    TCommand command_to_execute;
```

Comment:
I want to avoid unnecessary constructor/destructor calls, so it’s better to put it out of the while loop.

> On Jun 10, 2015, at 20:07, Dima <notifications@github.com> wrote:
> 
> In src/messages_queue.h <https://github.com/KnowSheet/Alohalytics/pull/36#discussion_r32140857>:
> 
> > +        // Also delete successfully processed archive.
> > +        std::remove(full_path_to_file.c_str());
> > +        return true;
> > +      } else {
> > +        result = ProcessingResult::EProcessingError;
> > +        // Stop processing archives on error.
> > +        return false;
> > +      }
> > +    });
> > +    if (callback) {
> > +      callback(result);
> > +    }
> > +  }
> > +
> > +  void WorkerThread() {
> > +    TCommand command_to_execute;
> This one can be moved into the while(true) loop.
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/KnowSheet/Alohalytics/pull/36/files#r32140857>.
>